File: src/agents/mohnish_pabrai.py
from src.graph.state import AgentState, show_agent_reasoning
from src.tools.api import get_financial_metrics, get_market_cap, search_line_items
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.messages import HumanMessage
from pydantic import BaseModel
import json
from typing_extensions import Literal
from src.utils.progress import progress
from src.utils.llm import call_llm
from src.utils.api_key import get_api_key_from_state


class MohnishPabraiSignal(BaseModel):
    signal: Literal["bullish", "bearish", "neutral"]
    confidence: float
    reasoning: str


def mohnish_pabrai_agent(state: AgentState, agent_id: str = "mohnish_pabrai_agent"):
    """
    Analyzes stocks using Mohnish Pabrai's principles:
    - Focus on undervalued, high-quality businesses.
    - Evaluate downside protection and cash yield.
    - Use growth, valuation, fundamentals, and insider activity signals.
    """
    data = state["data"]
    tickers = data["tickers"]
    end_date = data["end_date"]
    api_key = get_api_key_from_state(state, "FINANCIAL_DATASETS_API_KEY")

    analysis_data = {}
    pabrai_analysis = {}

    for ticker in tickers:
        progress.update_status(agent_id, ticker, "Fetching financial metrics")
        metrics = get_financial_metrics(ticker, end_date, period="annual", limit=8, api_key=api_key)

        progress.update_status(agent_id, ticker, "Gathering financial line items")
        # Safe line items extraction
        keywords = [
            "revenue", "gross_profit", "gross_margin", "operating_income", "operating_margin",
            "net_income", "free_cash_flow", "total_debt", "cash_and_equivalents",
            "current_assets", "current_liabilities", "shareholders_equity",
            "capital_expenditure", "depreciation_and_amortization", "outstanding_shares",
        ]
        financial_line_items = {}
        for kw in keywords:
            items = search_line_items(metrics, kw)
            if isinstance(items, dict):
                financial_line_items.update(items)
        
        # Analyze valuation safely
        pabrai_analysis[ticker] = analyze_pabrai_valuation(financial_line_items, metrics)

        # Analyze downside protection safely
        downside = analyze_downside_protection(financial_line_items)

        # Example growth analysis (stub, replace with your growth logic)
        growth_analysis = {"score": 5.0}  # placeholder for real growth scoring

        # Example insider activity (stub)
        insider_activity = {"score": 5.0}  # placeholder for real insider scoring

        # Combine partial scores with Pabrai-style weights
        total_score = (
            pabrai_analysis[ticker].get("score", 0) * 0.40
            + growth_analysis.get("score", 0) * 0.30
            + downside.get("score", 0) * 0.20
            + insider_activity.get("score", 0) * 0.10
        )
        max_possible_score = 10.0

        # Map total_score to signal
        if total_score >= 7.5:
            signal = "bullish"
        elif total_score <= 4.5:
            signal = "bearish"
        else:
            signal = "neutral"

        analysis_data[ticker] = {
            "signal": signal,
            "score": total_score,
            "max_score": max_possible_score,
            "valuation_analysis": pabrai_analysis[ticker],
            "downside_analysis": downside,
            "growth_analysis": growth_analysis,
            "insider_activity": insider_activity,
        }

    return {
    "data": analysis_data
}



def analyze_downside_protection(financial_line_items: list) -> dict[str, any]:
    """Assess balance-sheet strength and downside resiliency (capital preservation first)."""
    if not financial_line_items:
        return {"score": 0, "details": "Insufficient data"}

    latest = financial_line_items[0]
    details: list[str] = []
    score = 0

    cash = getattr(latest, "cash_and_equivalents", None)
    debt = getattr(latest, "total_debt", None)
    current_assets = getattr(latest, "current_assets", None)
    current_liabilities = getattr(latest, "current_liabilities", None)
    equity = getattr(latest, "shareholders_equity", None)

    # Net cash position is a strong downside protector
    net_cash = None
    if cash is not None and debt is not None:
        net_cash = cash - debt
        if net_cash > 0:
            score += 3
            details.append(f"Net cash position: ${net_cash:,.0f}")
        else:
            details.append(f"Net debt position: ${net_cash:,.0f}")

    # Current ratio
    if current_assets is not None and current_liabilities is not None and current_liabilities > 0:
        current_ratio = current_assets / current_liabilities
        if current_ratio >= 2.0:
            score += 2
            details.append(f"Strong liquidity (current ratio {current_ratio:.2f})")
        elif current_ratio >= 1.2:
            score += 1
            details.append(f"Adequate liquidity (current ratio {current_ratio:.2f})")
        else:
            details.append(f"Weak liquidity (current ratio {current_ratio:.2f})")

    # Low leverage
    if equity is not None and equity > 0 and debt is not None:
        de_ratio = debt / equity
        if de_ratio < 0.3:
            score += 2
            details.append(f"Very low leverage (D/E {de_ratio:.2f})")
        elif de_ratio < 0.7:
            score += 1
            details.append(f"Moderate leverage (D/E {de_ratio:.2f})")
        else:
            details.append(f"High leverage (D/E {de_ratio:.2f})")

    # Free cash flow positive and stable
    fcf_values = [getattr(li, "free_cash_flow", None) for li in financial_line_items if getattr(li, "free_cash_flow", None) is not None]
    if fcf_values and len(fcf_values) >= 3:
        recent_avg = sum(fcf_values[:3]) / 3
        older = sum(fcf_values[-3:]) / 3 if len(fcf_values) >= 6 else fcf_values[-1]
        if recent_avg > 0 and recent_avg >= older:
            score += 2
            details.append("Positive and improving/stable FCF")
        elif recent_avg > 0:
            score += 1
            details.append("Positive but declining FCF")
        else:
            details.append("Negative FCF")

    return {"score": min(10, score), "details": "; ".join(details)}


def analyze_pabrai_valuation(financials: dict, financial_line_items: dict) -> dict:
    """
    Analyze a company's valuation and fundamentals for Mohnish Pabrai's investing style.
    
    Args:
        financials (dict): Raw financial metrics data.
        financial_line_items (dict): Extracted line items relevant to valuation.
        
    Returns:
        dict: Contains a 'score' (0–10) and textual 'reasoning'.
    """
    def get_numeric(value):
        """Safely extract a numeric value from raw input or nested dict."""
        if isinstance(value, dict):
            # Try common numeric keys
            for k in ("value", "amount", "numeric"):
                if k in value and isinstance(value[k], (int, float)):
                    return value[k]
            return None
        elif isinstance(value, (int, float)):
            return value
        try:
            return float(value)
        except (TypeError, ValueError):
            return None

    # Extract key metrics safely
    market_cap = get_numeric(financial_line_items.get("market_cap"))
    pe_ratio = get_numeric(financial_line_items.get("pe_ratio"))
    debt = get_numeric(financial_line_items.get("total_debt"))
    cash = get_numeric(financial_line_items.get("cash_and_equivalents"))
    free_cash_flow = get_numeric(financial_line_items.get("free_cash_flow"))

    if not financial_line_items or market_cap is None or market_cap <= 0:
        return {"score": 0, "reasoning": "Insufficient financial data for valuation."}

    score = 0
    reasoning_parts = []

    # Simple scoring examples
    if pe_ratio is not None and pe_ratio < 15:
        score += 3
        reasoning_parts.append(f"Reasonable P/E ratio ({pe_ratio})")
    elif pe_ratio is not None:
        score += 1
        reasoning_parts.append(f"High P/E ratio ({pe_ratio})")

    if debt is not None and debt / market_cap < 0.5:
        score += 3
        reasoning_parts.append(f"Low debt relative to market cap ({debt}/{market_cap})")
    elif debt is not None:
        score += 1
        reasoning_parts.append(f"High debt relative to market cap ({debt}/{market_cap})")

    if free_cash_flow is not None and free_cash_flow > 0:
        score += 2
        reasoning_parts.append(f"Positive free cash flow ({free_cash_flow})")
    else:
        reasoning_parts.append("No positive free cash flow data")

    if cash is not None and cash / market_cap > 0.1:
        score += 2
        reasoning_parts.append(f"Healthy cash reserves relative to market cap ({cash}/{market_cap})")

    # Normalize score to 10
    if score > 10:
        score = 10

    return {
        "score": score,
        "reasoning": "; ".join(reasoning_parts),
        "market_cap": market_cap,
        "pe_ratio": pe_ratio,
        "debt": debt,
        "cash": cash,
        "free_cash_flow": free_cash_flow,
    }



def analyze_double_potential(financial_line_items: list, market_cap: float | None) -> dict[str, any]:
    """Estimate low-risk path to double capital in ~2-3 years: runway from FCF growth + rerating."""
    if not financial_line_items or market_cap is None or market_cap <= 0:
        return {"score": 0, "details": "Insufficient data"}

    details: list[str] = []

    # Use revenue and FCF trends as rough growth proxy (keep it simple)
    revenues = [getattr(li, "revenue", None) for li in financial_line_items if getattr(li, "revenue", None) is not None]
    fcfs = [getattr(li, "free_cash_flow", None) for li in financial_line_items if getattr(li, "free_cash_flow", None) is not None]

    score = 0
    if revenues and len(revenues) >= 3:
        recent_rev = sum(revenues[:3]) / 3
        older_rev = sum(revenues[-3:]) / 3 if len(revenues) >= 6 else revenues[-1]
        if older_rev > 0:
            rev_growth = (recent_rev / older_rev) - 1
            if rev_growth > 0.15:
                score += 2
                details.append(f"Strong revenue trajectory ({rev_growth:.1%})")
            elif rev_growth > 0.05:
                score += 1
                details.append(f"Modest revenue growth ({rev_growth:.1%})")

    if fcfs and len(fcfs) >= 3:
        recent_fcf = sum(fcfs[:3]) / 3
        older_fcf = sum(fcfs[-3:]) / 3 if len(fcfs) >= 6 else fcfs[-1]
        if older_fcf != 0:
            fcf_growth = (recent_fcf / older_fcf) - 1
            if fcf_growth > 0.20:
                score += 3
                details.append(f"Strong FCF growth ({fcf_growth:.1%})")
            elif fcf_growth > 0.08:
                score += 2
                details.append(f"Healthy FCF growth ({fcf_growth:.1%})")
            elif fcf_growth > 0:
                score += 1
                details.append(f"Positive FCF growth ({fcf_growth:.1%})")

    # If FCF yield is already high (>8%), doubling can come from cash generation alone in few years
    tmp_val = analyze_pabrai_valuation(financial_line_items, market_cap)
    fcf_yield = tmp_val.get("fcf_yield")
    if fcf_yield is not None:
        if fcf_yield > 0.08:
            score += 3
            details.append("High FCF yield can drive doubling via retained cash/Buybacks")
        elif fcf_yield > 0.05:
            score += 1
            details.append("Reasonable FCF yield supports moderate compounding")

    return {"score": min(10, score), "details": "; ".join(details)}


def generate_pabrai_output(
    ticker: str,
    analysis_data: dict[str, any],
    state: AgentState,
    agent_id: str,
) -> MohnishPabraiSignal:
    """Generate Pabrai-style decision focusing on low risk, high uncertainty bets and cloning."""
    template = ChatPromptTemplate.from_messages([
        (
          "system",
          """You are Mohnish Pabrai. Apply my value investing philosophy:

          - Heads I win; tails I don't lose much: prioritize downside protection first.
          - Buy businesses with simple, understandable models and durable moats.
          - Demand high free cash flow yields and low leverage; prefer asset-light models.
          - Look for situations where intrinsic value is rising and price is significantly lower.
          - Favor cloning great investors' ideas and checklists over novelty.
          - Seek potential to double capital in 2-3 years with low risk.
          - Avoid leverage, complexity, and fragile balance sheets.

            Provide candid, checklist-driven reasoning, with emphasis on capital preservation and expected mispricing.
            """,
        ),
        (
          "human",
          """Analyze {ticker} using the provided data.

          DATA:
          {analysis_data}

          Return EXACTLY this JSON:
          {{
            "signal": "bullish" | "bearish" | "neutral",
            "confidence": float (0-100),
            "reasoning": "string with Pabrai-style analysis focusing on downside protection, FCF yield, and doubling potential"
          }}
          """,
        ),
    ])

    prompt = template.invoke({
        "analysis_data": json.dumps(analysis_data, indent=2),
        "ticker": ticker,
    })

    def create_default_pabrai_signal():
        return MohnishPabraiSignal(signal="neutral", confidence=0.0, reasoning="Error in analysis, defaulting to neutral")

    return call_llm(
        prompt=prompt,
        state=state,
        pydantic_model=MohnishPabraiSignal,
        agent_name=agent_id,
        default_factory=create_default_pabrai_signal,
    ) 

File: src/agents/mohnish_pabrai.rs
// Source: src/agents/mohnish_pabrai.py
//! Sibling to src/agents/mohnish_pabrai.py
//! Analyzes stocks using Mohnish Pabrai's Dhandho investment principles.

use crate::data::models::LineItem;
use crate::graph::state::AgentState;
use crate::tools::api::{get_financial_metrics, get_market_cap, search_line_items};
use crate::utils::llm::call_llm;
use anyhow::{Context, Result};

#[derive(Debug, serde::Serialize, serde::Deserialize, Default)]
pub struct MohnishPabraiSignal {
    pub signal: String,  // "bullish" | "bearish" | "neutral"
    pub confidence: u32, // 0-100
    pub reasoning: String,
}

pub async fn mohnish_pabrai_agent(state: &mut AgentState, agent_id: &str) -> Result<()> {
    println!("Running Mohnish Pabrai Agent: {}", agent_id);

    let _start_date = state
        .data
        .get("start_date")
        .and_then(|v| v.as_str())
        .context("Missing start_date in mohnish_pabrai_agent")?;

    let end_date = state
        .data
        .get("end_date")
        .and_then(|v| v.as_str())
        .context("Missing end_date in mohnish_pabrai_agent")?;

    let tickers_json = state
        .data
        .get("tickers")
        .context("Missing tickers in mohnish_pabrai_agent")?;
    let tickers: Vec<String> = serde_json::from_value(tickers_json.clone())?;

    let api_key = state
        .metadata
        .get("FINANCIAL_DATASETS_API_KEY")
        .and_then(|v| v.as_str());

    let mut pabrai_analysis = serde_json::Map::new();

    for ticker in &tickers {
        // Fetch raw data
        let _metrics = get_financial_metrics(ticker, end_date, "annual", 8, api_key)
            .await
            .unwrap_or_default();
        let line_items = search_line_items(
            ticker,
            vec![
                "revenue".to_string(),
                "gross_profit".to_string(),
                "gross_margin".to_string(),
                "operating_income".to_string(),
                "operating_margin".to_string(),
                "net_income".to_string(),
                "free_cash_flow".to_string(),
                "total_debt".to_string(),
                "cash_and_equivalents".to_string(),
                "current_assets".to_string(),
                "current_liabilities".to_string(),
                "shareholders_equity".to_string(),
                "capital_expenditure".to_string(),
                "depreciation_and_amortization".to_string(),
                "outstanding_shares".to_string(),
            ],
            end_date,
            "annual",
            8,
            api_key,
        )
        .await
        .unwrap_or_default();

        let market_cap = get_market_cap(ticker, end_date, api_key)
            .await
            .unwrap_or(None)
            .unwrap_or(0.0);

        // Sub-analyses
        let downside = analyze_downside_protection(&line_items);
        let valuation = analyze_pabrai_valuation(&line_items, market_cap);
        let double_potential = analyze_double_potential(&line_items, market_cap);

        let total_score = downside.score as f64 * 0.45
            + valuation.score as f64 * 0.35
            + double_potential.score as f64 * 0.20;

        let max_score = 10.0;

        let signal = if total_score >= 7.5 {
            "bullish"
        } else if total_score <= 4.0 {
            "bearish"
        } else {
            "neutral"
        };

        let facts = serde_json::json!({
            "signal": signal,
            "score": (total_score * 100.0).round() / 100.0,
            "max_score": max_score,
            "downside_protection": {
                "score": downside.score,
                "details": downside.details,
            },
            "valuation": {
                "score": valuation.score,
                "details": valuation.details,
                "fcf_yield": valuation.fcf_yield,
                "normalized_fcf": valuation.normalized_fcf,
            },
            "double_potential": {
                "score": double_potential.score,
                "details": double_potential.details,
            },
            "market_cap": market_cap,
        });

        let output = generate_pabrai_output(ticker, &facts, state, agent_id).await?;

        pabrai_analysis.insert(
            ticker.clone(),
            serde_json::json!({
                "signal": output.signal,
                "confidence": output.confidence,
                "reasoning": output.reasoning,
            }),
        );
    }

    let analyst_signals = state
        .data
        .entry("analyst_signals".to_string())
        .or_insert_with(|| serde_json::json!({}));
    if let Some(obj) = analyst_signals.as_object_mut() {
        obj.insert(
            agent_id.to_string(),
            serde_json::Value::Object(pabrai_analysis),
        );
    }

    Ok(())
}

pub struct PabraiSubResult {
    pub score: u32,
    pub details: String,
}

pub fn analyze_downside_protection(financial_line_items: &[LineItem]) -> PabraiSubResult {
    let mut score = 0;
    let mut details = Vec::new();

    if financial_line_items.is_empty() {
        return PabraiSubResult {
            score: 0,
            details: "Insufficient data".to_string(),
        };
    }

    let latest = &financial_line_items[0];
    let cash = latest.cash_and_equivalents.unwrap_or(0.0);
    let debt = latest.total_debt.unwrap_or(0.0);
    let current_assets = latest.current_assets.unwrap_or(0.0);
    let current_liabilities = latest.current_liabilities.unwrap_or(0.0);
    let equity = latest.shareholders_equity.unwrap_or(0.0);

    // 1. Net Cash
    let net_cash = cash - debt;
    if net_cash > 0.0 {
        score += 3;
        details.push(format!("Net cash position: ${:.0}", net_cash));
    } else {
        details.push(format!("Net debt position: ${:.0}", net_cash));
    }

    // 2. Current ratio
    if current_liabilities > 0.0 {
        let current_ratio = current_assets / current_liabilities;
        if current_ratio >= 2.0 {
            score += 2;
            details.push(format!(
                "Strong liquidity (current ratio {:.2})",
                current_ratio
            ));
        } else if current_ratio >= 1.2 {
            score += 1;
            details.push(format!(
                "Adequate liquidity (current ratio {:.2})",
                current_ratio
            ));
        } else {
            details.push(format!(
                "Weak liquidity (current ratio {:.2})",
                current_ratio
            ));
        }
    }

    // 3. Leverage
    if equity > 0.0 {
        let de_ratio = debt / equity;
        if de_ratio < 0.3 {
            score += 2;
            details.push(format!("Very low leverage (D/E {:.2})", de_ratio));
        } else if de_ratio < 0.7 {
            score += 1;
            details.push(format!("Moderate leverage (D/E {:.2})", de_ratio));
        } else {
            details.push(format!("High leverage (D/E {:.2})", de_ratio));
        }
    }

    // 4. Stable/improving FCF
    let fcf_values: Vec<f64> = financial_line_items
        .iter()
        .filter_map(|li| li.free_cash_flow)
        .collect();
    if fcf_values.len() >= 3 {
        let recent_avg: f64 = fcf_values[..3].iter().sum::<f64>() / 3.0;
        let older = if fcf_values.len() >= 6 {
            fcf_values[fcf_values.len() - 3..].iter().sum::<f64>() / 3.0
        } else {
            fcf_values[fcf_values.len() - 1]
        };

        if recent_avg > 0.0 && recent_avg >= older {
            score += 2;
            details.push("Positive and improving/stable FCF".to_string());
        } else if recent_avg > 0.0 {
            score += 1;
            details.push("Positive but declining FCF".to_string());
        } else {
            details.push("Negative FCF".to_string());
        }
    }

    PabraiSubResult {
        score: score.min(10),
        details: details.join("; "),
    }
}

pub struct PabraiValResult {
    pub score: u32,
    pub details: String,
    pub fcf_yield: Option<f64>,
    pub normalized_fcf: Option<f64>,
}

pub fn analyze_pabrai_valuation(
    financial_line_items: &[LineItem],
    market_cap: f64,
) -> PabraiValResult {
    if financial_line_items.is_empty() || market_cap <= 0.0 {
        return PabraiValResult {
            score: 0,
            details: "Insufficient data".to_string(),
            fcf_yield: None,
            normalized_fcf: None,
        };
    }

    let mut details = Vec::new();
    let fcf_values: Vec<f64> = financial_line_items
        .iter()
        .filter_map(|li| li.free_cash_flow)
        .collect();

    if fcf_values.len() < 3 {
        return PabraiValResult {
            score: 0,
            details: "Insufficient FCF history".to_string(),
            fcf_yield: None,
            normalized_fcf: None,
        };
    }

    let periods = 5.min(fcf_values.len());
    let normalized_fcf: f64 = fcf_values[..periods].iter().sum::<f64>() / periods as f64;
    if normalized_fcf <= 0.0 {
        return PabraiValResult {
            score: 0,
            details: "Non-positive normalized FCF".to_string(),
            fcf_yield: None,
            normalized_fcf: Some(normalized_fcf),
        };
    }

    let fcf_yield = normalized_fcf / market_cap;
    let mut score = 0;

    if fcf_yield > 0.10 {
        score += 4;
        details.push(format!(
            "Exceptional value: {:.1}% FCF yield",
            fcf_yield * 100.0
        ));
    } else if fcf_yield > 0.07 {
        score += 3;
        details.push(format!(
            "Attractive value: {:.1}% FCF yield",
            fcf_yield * 100.0
        ));
    } else if fcf_yield > 0.05 {
        score += 2;
        details.push(format!(
            "Reasonable value: {:.1}% FCF yield",
            fcf_yield * 100.0
        ));
    } else if fcf_yield > 0.03 {
        score += 1;
        details.push(format!(
            "Borderline value: {:.1}% FCF yield",
            fcf_yield * 100.0
        ));
    } else {
        details.push(format!("Expensive: {:.1}% FCF yield", fcf_yield * 100.0));
    }

    // Asset-light tilt (Avg capex < 5% of revenue (+2) or < 10% (+1))
    let mut capex_to_revenue = Vec::new();
    for item in financial_line_items {
        if let (Some(rev), Some(capex)) = (item.revenue, item.capital_expenditure) {
            if rev > 0.0 {
                capex_to_revenue.push(capex.abs() / rev);
            }
        }
    }
    if !capex_to_revenue.is_empty() {
        let avg_ratio: f64 = capex_to_revenue.iter().sum::<f64>() / capex_to_revenue.len() as f64;
        if avg_ratio < 0.05 {
            score += 2;
            details.push(format!(
                "Asset-light: Avg capex {:.1}% of revenue",
                avg_ratio * 100.0
            ));
        } else if avg_ratio < 0.10 {
            score += 1;
            details.push(format!(
                "Moderate capex: Avg capex {:.1}% of revenue",
                avg_ratio * 100.0
            ));
        } else {
            details.push(format!(
                "Capex heavy: Avg capex {:.1}% of revenue",
                avg_ratio * 100.0
            ));
        }
    }

    PabraiValResult {
        score: score.min(10),
        details: details.join("; "),
        fcf_yield: Some(fcf_yield),
        normalized_fcf: Some(normalized_fcf),
    }
}

pub fn analyze_double_potential(
    financial_line_items: &[LineItem],
    market_cap: f64,
) -> PabraiSubResult {
    let mut score = 0;
    let mut details = Vec::new();

    if financial_line_items.is_empty() || market_cap <= 0.0 {
        return PabraiSubResult {
            score: 0,
            details: "Insufficient data".to_string(),
        };
    }

    let revenues: Vec<f64> = financial_line_items
        .iter()
        .filter_map(|li| li.revenue)
        .collect();
    let fcfs: Vec<f64> = financial_line_items
        .iter()
        .filter_map(|li| li.free_cash_flow)
        .collect();

    // 1. Revenue CAGR/Growth proxy
    if revenues.len() >= 3 {
        let recent_rev: f64 = revenues[..3].iter().sum::<f64>() / 3.0;
        let older_rev = if revenues.len() >= 6 {
            revenues[revenues.len() - 3..].iter().sum::<f64>() / 3.0
        } else {
            revenues[revenues.len() - 1]
        };
        if older_rev > 0.0 {
            let rev_growth = (recent_rev / older_rev) - 1.0;
            if rev_growth > 0.15 {
                score += 2;
                details.push(format!(
                    "Strong revenue trajectory ({:.1}%)",
                    rev_growth * 100.0
                ));
            } else if rev_growth > 0.05 {
                score += 1;
                details.push(format!(
                    "Modest revenue growth ({:.1}%)",
                    rev_growth * 100.0
                ));
            }
        }
    }

    // 2. FCF growth proxy
    if fcfs.len() >= 3 {
        let recent_fcf: f64 = fcfs[..3].iter().sum::<f64>() / 3.0;
        let older_fcf = if fcfs.len() >= 6 {
            fcfs[fcfs.len() - 3..].iter().sum::<f64>() / 3.0
        } else {
            fcfs[fcfs.len() - 1]
        };
        if older_fcf != 0.0 {
            let fcf_growth = (recent_fcf / older_fcf) - 1.0;
            if fcf_growth > 0.20 {
                score += 3;
                details.push(format!("Strong FCF growth ({:.1}%)", fcf_growth * 100.0));
            } else if fcf_growth > 0.08 {
                score += 2;
                details.push(format!("Healthy FCF growth ({:.1}%)", fcf_growth * 100.0));
            } else if fcf_growth > 0.0 {
                score += 1;
                details.push(format!("Positive FCF growth ({:.1}%)", fcf_growth * 100.0));
            }
        }
    }

    // 3. FCF yield compounding check
    let val_tmp = analyze_pabrai_valuation(financial_line_items, market_cap);
    if let Some(fcf_yield) = val_tmp.fcf_yield {
        if fcf_yield > 0.08 {
            score += 3;
            details
                .push("High FCF yield can drive doubling via retained cash/buybacks".to_string());
        } else if fcf_yield > 0.05 {
            score += 1;
            details.push("Reasonable FCF yield supports moderate compounding".to_string());
        }
    }

    PabraiSubResult {
        score: score.min(10),
        details: details.join("; "),
    }
}

pub async fn generate_pabrai_output(
    ticker: &str,
    facts: &serde_json::Value,
    state: &AgentState,
    agent_id: &str,
) -> Result<MohnishPabraiSignal> {
    let system_prompt = "You are Mohnish Pabrai. Apply my value investing philosophy:\n\
        - Heads I win; tails I don't lose much: prioritize downside protection first.\n\
        - Buy businesses with simple, understandable models and durable moats.\n\
        - Demand high free cash flow yields and low leverage; prefer asset-light models.\n\
        - Seek potential to double capital in 2-3 years with low risk.\n\
        - Avoid leverage, complexity, and fragile balance sheets.\n\
        Return JSON matching exactly: {\"signal\": \"bullish\" | \"bearish\" | \"neutral\", \"confidence\": int, \"reasoning\": \"short\"}";

    let user_prompt = format!(
        "Ticker: {}\nFacts:\n{}\n\nReturn structured JSON output.",
        ticker,
        serde_json::to_string_pretty(facts)?
    );

    call_llm(system_prompt, &user_prompt, Some(agent_id), Some(state), 3).await
}
